engine/flink/util/src/main/scala/pl/touk/nussknacker/engine/flink/util/keyed.scala

```diff
@@ -49,40 +49,48 @@ object keyed {
   }
 
   /* TODO, FIXME: Errors on interpret should be handled with FlinkEspExceptionHandler; see KeyedRecordFlatMapper */
-  abstract class BaseKeyedValueMapper[OutputKey <: AnyRef: TypeTag] extends RichMapFunction[Context, ValueWithContext[KeyedValue[OutputKey, AnyRef]]] with LazyParameterInterpreterFunction {
+  abstract class BaseKeyedValueMapper[OutputKey <: AnyRef: TypeTag, OutputValue <:AnyRef: TypeTag] extends RichMapFunction[Context, ValueWithContext[KeyedValue[OutputKey, OutputValue]]] with LazyParameterInterpreterFunction {
 
     protected implicit def lazyParameterInterpreterImpl: LazyParameterInterpreter = lazyParameterInterpreter
 
-    protected def prepareInterpreter(key: LazyParameter[OutputKey], value: LazyParameter[AnyRef]): Context => KeyedValue[OutputKey, AnyRef] = {
+    protected def prepareInterpreter(key: LazyParameter[OutputKey], value: LazyParameter[OutputValue]): Context => KeyedValue[OutputKey, OutputValue] = {
       lazyParameterInterpreter.syncInterpretationFunction(
         key.product(value).map(tuple => KeyedValue(tuple._1, tuple._2)))
     }
 
-    protected def interpret(ctx: Context): KeyedValue[OutputKey, AnyRef]
+    protected def interpret(ctx: Context): KeyedValue[OutputKey, OutputValue]
 
-    override def map(ctx: Context): ValueWithContext[KeyedValue[OutputKey, AnyRef]] = ValueWithContext(interpret(ctx), ctx)
+    override def map(ctx: Context): ValueWithContext[KeyedValue[OutputKey, OutputValue]] = ValueWithContext(interpret(ctx), ctx)
 
   }
 
   class KeyedValueMapper(protected val lazyParameterHelper: FlinkLazyParameterFunctionHelper, key: LazyParameter[AnyRef], value: LazyParameter[AnyRef])
-    extends BaseKeyedValueMapper[AnyRef] {
+    extends BaseKeyedValueMapper[AnyRef, AnyRef] {
 
     private lazy val interpreter = prepareInterpreter(key, value)
 
     override protected def interpret(ctx: Context): KeyedValue[AnyRef, AnyRef] = interpreter(ctx)
 
   }
 
-  class StringKeyedValueMapper(protected val lazyParameterHelper: FlinkLazyParameterFunctionHelper, key: LazyParameter[CharSequence], value: LazyParameter[AnyRef])
-    extends BaseKeyedValueMapper[String] {
 
-    private lazy val interpreter = prepareInterpreter(key.map(transformKey), value)
+
+  class StringKeyedValueMapper[T<:AnyRef:TypeTag](protected val lazyParameterHelper: FlinkLazyParameterFunctionHelper,
+                                                  key: LazyParameter[CharSequence],
+                                                  value: LazyParameterInterpreter => LazyParameter[T])
```

Comment:
Parameter that is a function taking parameter interpreter and returning lazy parameter that will be interpreted by some other parameter interpreter? It looks like we need at least comment in this place ;)

Reply:
I added comment :)